Mathlib/Tactic/SolveByElim.lean

```diff
@@ -65,121 +329,197 @@ This would make it impossible to apply the lemma
 a second time with different values of the metavariables.
 
 See https://github.com/leanprover-community/mathlib/issues/2269
+
+### Relevant local hypotheses
+
+`solve_by_elim*` works with multiple goals,
+and we need to use separate sets of local hypotheses for each goal.
+The second component of the returned value provides these local hypotheses.
+(Essentially using `local_context`, along with some filtering to remove hypotheses
+that have been explicitly removed via `only` or `[-h]`.)
+
 -/
-def mkAssumptionSet (noDflt : Bool) (hs : List (TSyntax `term)) :
+-- These `TermElabM`s must be run inside a suitable `g.withContext`,
+-- usually using `elabContextLemmas`.
+def mkAssumptionSet (noDefaults star : Bool) (add remove : List Term) :
     MetaM (List (TermElabM Expr) × TermElabM (List Expr)) := do
-  let hs := if noDflt then hs else [← `(rfl), ← `(trivial), ← `(congrFun), ← `(congrArg)] ++ hs
-  let hs := hs.map (λ s => Elab.Term.elabTerm s.raw none)
-  let locals : TermElabM (List Expr) := if noDflt then pure [] else pure (← getLocalHyps).toList
-  return (hs, locals)
+  if star && !noDefaults then
+    throwError "It does make sense to use `*` without `only`."
+
+  let (addLocal, addExpr) ← partitionLocalHyps add
+  let (removeLocal, removeExpr) ← partitionLocalHyps remove
+
+  if !removeExpr.isEmpty then
+    throwError "It doesn't make sense to remove expressions which are not local hypotheses."
+  let defaults : List Term := [← `(rfl), ← `(trivial), ← `(congrFun), ← `(congrArg)]
+  let hyps := (if noDefaults then addExpr else defaults ++ addExpr).map elab'
+
+  if !removeLocal.isEmpty && noDefaults && !star then
+    throwError "It doesn't make sense to remove local hypotheses when using `only` without `*`."
+  -- TODO Consider extracting `FVarId`s to avoid re-elaborating here.
+  let locals : TermElabM (List Expr) := if noDefaults && !star then do
+    pure <| (addLocal.removeAll removeLocal).map .fvar
+  else do
+    if !addLocal.isEmpty then
+      throwError "It doesn't make sense to add local hypotheses unless you use `only` without `*`."
+    pure <| (← getLocalHyps).toList.removeAll (removeLocal.map .fvar)
+
+  return (hyps, locals)
+  where
+  /-- Run `elabTerm`. -/
+  elab' (t : Term) : TermElabM Expr := Elab.Term.elabTerm t.raw none
+
+/-- Syntax for omitting a local hypothesis in `solve_by_elim`. -/
+syntax erase := "-" term:max
```

Comment:
Did you consider re-using the existing syntax from `simp` for this? (Maybe that would save some duplication, but maybe also it's cleaner to keep things stratified as you have here.)

Reply:
It is less general that the syntax for simp, as it doesn't allow the various left, up, and down arrows in front of lemma names. Since that syntax is in core, there isn't a convenient way to immediately refactor it to support this syntax.

Reply:
Maybe leave a note here and ask for addition of "-" in core for simp?